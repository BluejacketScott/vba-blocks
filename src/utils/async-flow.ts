// TODO Advanced typing to match input and output type from functions
// For now, use simple 1-to-1 mapping

type AsyncFn<TValue> = (input: TValue) => Promise<TValue>;
type Fn<TValue> = (input: TValue) => TValue;

export default function asyncMap<TValue>(
  ...fns: Array<Fn<TValue> | AsyncFn<TValue>>
): AsyncFn<TValue> {
  return value => fns.reduce(async (memo, fn) => fn(await memo), Promise.resolve(value));
}
