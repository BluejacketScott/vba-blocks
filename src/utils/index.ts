export { default as commaSeparated } from './comma-separated';
export { default as convertToToml } from './convert-to-toml';
export { default as download } from './download';
export { default as getStaging } from './get-staging';
export * from './fs';
export * from './git';
export { default as has } from './has';
export { default as hash } from './hash';
export * from './is';
export { default as isPackaged } from './is-packaged';
export { default as last } from './last';
export { default as nonce } from './nonce';
export { default as optionList } from './option-list';
export { default as parallel } from './parallel';
export { default as run, escape, RunResult } from './run';
export { default as unique } from './unique';
export { default as unixJoin } from './unix-join';
export { default as unixPath } from './unix-path';
export { default as without } from './without';
export * from './zip';
