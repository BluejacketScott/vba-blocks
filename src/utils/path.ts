import {
	join as _join,
	normalize as _normalize,
	relative as _relative,
	resolve as _resolve
} from "path";
import sanitizeFilename from "sanitize-filename";

export { basename, dirname, extname } from "path";

const WINDOWS_REGEX = /\\/g;
const LEADING_SLASH = "./";

export function join(...parts: string[]): string {
	return normalize(_join(...parts));
}

export function normalize(value: string): string {
	let normalized = _normalize(value).replace(WINDOWS_REGEX, "/");
	if (value.startsWith(LEADING_SLASH)) normalized = LEADING_SLASH + normalized;

	return normalized;
}

export function relative(from: string, to: string): string {
	return normalize(_relative(from, to));
}

export function resolve(...parts: string[]): string {
	return normalize(_resolve(...parts));
}

export function sanitize(name: string): string {
	return sanitizeFilename(name, { replacement: "-" });
}

export function trailing(value: string): string {
	return value.endsWith("/") ? value : `${value}/`;
}
