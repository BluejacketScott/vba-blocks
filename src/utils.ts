import * as fs from './utils/fs';
import * as git from './utils/git';
import * as github from './utils/github';
import * as localGit from './utils/local-git';
import * as path from './utils/path';
import * as s3 from './utils/s3';
import * as toml from './utils/toml';
import * as xml from './utils/xml';

export { default as asyncMap } from './utils/async-flow';
export { default as download } from './utils/download';
export { default as getStaging } from './utils/get-staging';
export { default as has } from './utils/has';
export { default as hash } from './utils/hash';
export * from './utils/is';
export { default as noop } from './utils/noop';
export { default as Observable } from './utils/observable';
export { default as parallel } from './utils/parallel';
export * from './utils/pipe';
export { default as run } from './utils/run';
export { default as stdoutFile, watchFile } from './utils/stdout-file';
export * from './utils/text';
export { default as unique } from './utils/unique';
export { default as without } from './utils/without';
export * from './utils/zip';
export { fs, git, github, localGit, path, toml, xml, s3 };
