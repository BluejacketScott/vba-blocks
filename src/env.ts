import envPaths from "env-paths";
import { Reporter, reporter } from "./reporter";
import getStaging from "./utils/get-staging";
import { join } from "./utils/path";

export interface Env {
	isWindows: boolean;
	cwd: string;
	values: any;

	data: string;
	config: string;
	cache: string;
	log: string;
	temp: string;

	addins: string;
	scripts: string;
	bin: string;
	registry: string;
	packages: string;
	sources: string;
	staging: string;

	reporter: Reporter;
	debug: (id: string) => (...values: any[]) => void;
	silent: boolean;
}

const paths = envPaths("vba-blocks", { suffix: "" });

const cache = paths.cache;
const root = join(__dirname, "../");

const env: Env = {
	isWindows: process.platform === "win32",
	cwd: process.cwd(),
	values: process.env,

	...paths, // data, config, cache, log, temp
	addins: join(root, "addins/build"),
	scripts: join(root, "run-scripts"),
	bin: join(root, "bin"),
	registry: join(cache, "registry"),
	packages: join(cache, "packages"),
	sources: join(cache, "sources"),
	staging: getStaging(paths.temp),

	reporter,
	debug: id => {
		// Late-bind debug to allow loading --debug first
		let debug: (...args: any[]) => any;

		return (...args) => {
			if (!debug) {
				const _debug = require("./debug").default;
				debug = _debug(id);
			}

			return debug(...args);
		};
	},
	silent: false
};

export default env;
