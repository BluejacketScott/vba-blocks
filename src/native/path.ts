import { normalize } from 'path';
import env from '../env';
import exec from './exec';
import { CliError } from '../errors';
import { homedir } from 'os';
import { join } from '../utils/path';
import { readFile, writeFile } from '../utils/fs';
import { pathExists } from 'fs-extra';

export async function addToPATH() {
  if (env.isWindows) {
    try {
      const bin = normalize(env.bin);
      await exec(['add-to-path', bin]);
    } catch (underlying) {
      throw new CliError('Failed to add vba-blocks to PATH.', { underlying });
    }
  } else {
    // Add bin to ~/.profile and ~/.bash_profile (if it exists)
    const profile = join(homedir(), '.profile');
    const bash = join(homedir(), '.bash_profile');
    const profiles = [profile];
    if (await pathExists(bash)) profiles.push(bash);

    await Promise.all(profiles.map(addToProfile));
  }
}

export async function removeFromPATH() {
  if (env.isWindows) {
    try {
      await exec(['remove-from-path', normalize(env.bin)]);
    } catch (underlying) {
      throw new CliError('Failed to remove vba-blocks from PATH.', {
        underlying
      });
    }
  } else {
    const profile = join(homedir(), '.profile');
    const bash = join(homedir(), '.bash_profile');
    const profiles = [profile];
    if (await pathExists(bash)) profiles.push(bash);

    await Promise.all(profiles.map(removeFromProfile));
  }
}

export async function addToProfile(profile: string) {
  const path = `export PATH="${env.bin}:$PATH"`;
  let file = (await pathExists(profile))
    ? await readFile(profile, 'utf-8')
    : '';

  if (file.includes(path)) return;

  file += `\n${path}`;
  await writeFile(profile, file);
}

export async function removeFromProfile(profile: string) {
  if (!(await pathExists(profile))) return;

  const path = `export PATH="${env.bin}:$PATH"`;

  let file = await readFile(profile, 'utf-8');
  const index = file.indexOf(path);

  if (index < 0) return;

  file = file.substr(0, index) + file.substr(index + path.length);
  await writeFile(profile, file);
}
