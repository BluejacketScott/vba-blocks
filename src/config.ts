import { join } from './utils/path';
import { pathExists, readFile } from './utils/fs';
import { parse as parseToml } from './utils/toml';
import env from './env';
import { Sources, RegistrySource, PathSource, GitSource } from './sources';

export type Registry = {} | { [name: string]: { index: string; packages: string } };

export interface Flags {}

export interface Config {
  registry: Registry;
  flags: Flags;
  sources: Sources;
}

export interface ConfigValue {
  registry?: Registry;
  flags?: Flags;
}

const empty: ConfigValue = { registry: {}, flags: {} };
const defaults: ConfigValue = {
  registry: {
    'vba-blocks': {
      index: 'https://github.com/vba-blocks/registry',
      packages: 'https://packages.vba-blocks.com'
    }
  },
  flags: {}
};

/**
 * Load config, from local, user, and environment values
 *
 * - Load ~/.vba-blocks/config.toml (user)
 * - Search for .vba-blocks/config.toml up from cwd (local)
 * - Load VBA_BLOCKS_* from environment (override)
 */
export async function loadConfig(): Promise<Config> {
  const user: ConfigValue = {
    ...empty,
    ...((await readConfig(env.cache)) || {})
  };

  const dir = await findConfig(env.cwd);
  const local: ConfigValue = {
    ...empty,
    ...(dir ? await readConfig(dir) : {})
  };

  const override = loadConfigFromEnv();

  const registry: Registry = {
    ...defaults.registry,
    ...user.registry,
    ...local.registry,
    ...override.registry
  };
  const flags: Flags = {
    ...defaults.flags,
    ...user.flags,
    ...local.flags,
    ...override.flags
  };

  const sources: Sources = {
    registry: {},
    git: new GitSource(),
    path: new PathSource()
  };

  for (const [name, { index, packages }] of Object.entries(registry)) {
    sources.registry[name] = new RegistrySource({
      name,
      index,
      packages
    });
  }

  return { registry, flags, sources };
}

// Read config from dir (if present)
export async function readConfig(dir: string): Promise<ConfigValue | undefined> {
  const file = join(dir, 'config.toml');
  if (!(await pathExists(file))) return {};

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());

  return parsed;
}

// Find config up from and including given dir
// (looking for .vba-blocks/config.toml)
export async function findConfig(dir: string): Promise<string | undefined> {
  // TODO Search for .vba-blocks/config.toml starting at cwd
  return;
}

// Load config values from environment
export function loadConfigFromEnv(): ConfigValue {
  // TODO Load override config from VBA_BLOCKS_* env variables
  return {};
}
