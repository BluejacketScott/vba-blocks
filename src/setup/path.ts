import { normalize } from 'path';
import env from '../env';
import exec from './exec';
import { CliError } from '../errors';

export async function addToPATH() {
  try {
    await exec(['add-to-path', normalize(env.bin)]);
  } catch (underlying) {
    throw new CliError('Failed to add vba-blocks to PATH.', { underlying });
  }
}

export async function removeFromPATH() {
  try {
    await exec(['remove-from-path', normalize(env.bin)]);
  } catch (underlying) {
    throw new CliError('Failed to remove vba-blocks from PATH.', {
      underlying
    });
  }
}
