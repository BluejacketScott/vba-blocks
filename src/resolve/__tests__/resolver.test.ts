import { cache } from '../../../tests/__fixtures__';
import { loadConfig } from '../../config';
import env from '../../env';
import { join } from '../../utils/path';
import Resolver from '../resolver';

jest.mock('../../utils/git');

const original_env = { ...env };

beforeEach(() => {
  env.cache = cache;
  env.registry = join(cache, 'registry');
  env.packages = join(cache, 'packages');
  env.sources = join(cache, 'sources');
});

afterEach(() => {
  env.cache = original_env.cache;
  env.registry = original_env.registry;
  env.packages = original_env.packages;
  env.sources = original_env.sources;
});

const dependency = {
  name: 'vba-tools/json',
  version: '^2',
  registry: 'vba-blocks'
};

test('should get dependency from clean state', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);

  const resolved = await resolver.get(dependency);
  expect(resolved).toMatchSnapshot();
});

test('should get dependency with existing', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);

  resolver.graph.set('vba-tools/json', { name: 'vba-tools/json', range: [], registered: [] });

  const resolved = await resolver.get(dependency);
  expect(resolved).toMatchSnapshot();
});

test('should include preferred with resolved', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);

  resolver.prefer([
    {
      name: 'vba-tools/json',
      id: 'vba-tools/json@2.0.1',
      source: 'registry+vba-blocks#<hash>',
      version: '2.0.1',
      dependencies: [
        {
          name: 'vba-tools/dictionary',
          version: '^1',
          registry: 'vba-blocks'
        }
      ]
    }
  ]);

  const resolved = await resolver.get(dependency);
  expect(resolved).toMatchSnapshot();
});

test('should get registration', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);

  await resolver.get(dependency);

  const registration = resolver.getRegistration('vba-tools/json@2.0.1');
  expect(registration).toMatchSnapshot();
});

test('should iterate through graph entries', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);

  await resolver.get(dependency);

  const graph = [...resolver];
  expect(graph).toMatchSnapshot();
});
