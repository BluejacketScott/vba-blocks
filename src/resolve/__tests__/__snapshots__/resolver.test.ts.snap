// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get dependency from clean state 1`] = `
Object {
  "name": "vba-tools/json",
  "range": Array [
    "^2",
  ],
  "registered": Array [
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@1.0.3",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-bfa82d85e9964ac5fd9dbb014319b84eed127b93fca5495fab0256ce22b28b38",
      "version": "1.0.3",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@2.0.1",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-88fe488b3062b8c734923afa76dbf6628044ea0bc9bb460b4a9b2e2dd514f28e",
      "version": "2.0.1",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@2.1.0",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-c00e834ea96c4a6968d167b6616b2fa44a13fb672ff690b60a13326ca18de711",
      "version": "2.1.0",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@2.2.4",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-d788c54fff7e78fe95ced0684ebda35a17150751d9f629c45d2e705d6d9b6a62",
      "version": "2.2.4",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@2.3.0",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-3c2760ce2940dd19c16542cead792e0d05e58566155b1fbc72d445fe50a36799",
      "version": "2.3.0",
    },
  ],
}
`;

exports[`should get dependency with existing 1`] = `
Object {
  "name": "vba-tools/json",
  "range": Array [
    "^2",
  ],
  "registered": Array [],
}
`;

exports[`should get registration 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "vba-tools/dictionary",
      "registry": "vba-blocks",
      "version": "^1",
    },
  ],
  "id": "vba-tools/json@2.0.1",
  "name": "vba-tools/json",
  "source": "registry+vba-blocks#sha256-88fe488b3062b8c734923afa76dbf6628044ea0bc9bb460b4a9b2e2dd514f28e",
  "version": "2.0.1",
}
`;

exports[`should include preferred with resolved 1`] = `
Object {
  "name": "vba-tools/json",
  "preferred": Object {
    "dependencies": Array [
      Object {
        "name": "vba-tools/dictionary",
        "registry": "vba-blocks",
        "version": "^1",
      },
    ],
    "id": "vba-tools/json@2.0.1",
    "name": "vba-tools/json",
    "source": "registry+vba-blocks#<hash>",
    "version": "2.0.1",
  },
  "range": Array [
    "^2",
  ],
  "registered": Array [
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@1.0.3",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-bfa82d85e9964ac5fd9dbb014319b84eed127b93fca5495fab0256ce22b28b38",
      "version": "1.0.3",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@2.0.1",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-88fe488b3062b8c734923afa76dbf6628044ea0bc9bb460b4a9b2e2dd514f28e",
      "version": "2.0.1",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@2.1.0",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-c00e834ea96c4a6968d167b6616b2fa44a13fb672ff690b60a13326ca18de711",
      "version": "2.1.0",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@2.2.4",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-d788c54fff7e78fe95ced0684ebda35a17150751d9f629c45d2e705d6d9b6a62",
      "version": "2.2.4",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "vba-tools/dictionary",
          "registry": "vba-blocks",
          "version": "^1",
        },
      ],
      "id": "vba-tools/json@2.3.0",
      "name": "vba-tools/json",
      "source": "registry+vba-blocks#sha256-3c2760ce2940dd19c16542cead792e0d05e58566155b1fbc72d445fe50a36799",
      "version": "2.3.0",
    },
  ],
}
`;

exports[`should iterate through graph entries 1`] = `
Array [
  Array [
    "vba-tools/json",
    Object {
      "name": "vba-tools/json",
      "range": Array [
        "^2",
      ],
      "registered": Array [
        Object {
          "dependencies": Array [
            Object {
              "name": "vba-tools/dictionary",
              "registry": "vba-blocks",
              "version": "^1",
            },
          ],
          "id": "vba-tools/json@1.0.3",
          "name": "vba-tools/json",
          "source": "registry+vba-blocks#sha256-bfa82d85e9964ac5fd9dbb014319b84eed127b93fca5495fab0256ce22b28b38",
          "version": "1.0.3",
        },
        Object {
          "dependencies": Array [
            Object {
              "name": "vba-tools/dictionary",
              "registry": "vba-blocks",
              "version": "^1",
            },
          ],
          "id": "vba-tools/json@2.0.1",
          "name": "vba-tools/json",
          "source": "registry+vba-blocks#sha256-88fe488b3062b8c734923afa76dbf6628044ea0bc9bb460b4a9b2e2dd514f28e",
          "version": "2.0.1",
        },
        Object {
          "dependencies": Array [
            Object {
              "name": "vba-tools/dictionary",
              "registry": "vba-blocks",
              "version": "^1",
            },
          ],
          "id": "vba-tools/json@2.1.0",
          "name": "vba-tools/json",
          "source": "registry+vba-blocks#sha256-c00e834ea96c4a6968d167b6616b2fa44a13fb672ff690b60a13326ca18de711",
          "version": "2.1.0",
        },
        Object {
          "dependencies": Array [
            Object {
              "name": "vba-tools/dictionary",
              "registry": "vba-blocks",
              "version": "^1",
            },
          ],
          "id": "vba-tools/json@2.2.4",
          "name": "vba-tools/json",
          "source": "registry+vba-blocks#sha256-d788c54fff7e78fe95ced0684ebda35a17150751d9f629c45d2e705d6d9b6a62",
          "version": "2.2.4",
        },
        Object {
          "dependencies": Array [
            Object {
              "name": "vba-tools/dictionary",
              "registry": "vba-blocks",
              "version": "^1",
            },
          ],
          "id": "vba-tools/json@2.3.0",
          "name": "vba-tools/json",
          "source": "registry+vba-blocks#sha256-3c2760ce2940dd19c16542cead792e0d05e58566155b1fbc72d445fe50a36799",
          "version": "2.3.0",
        },
      ],
    },
  ],
]
`;
